### ISSUE TITLE
Request to recognize and color interior holes in parts 

### ISSUE THREAD

**Issue by marmilicious** [Jack000/Deepnest#5]
Typically in laser cutting we want to cut out all the interior lines (holes in a part) before cutting around the perimeter of a part.  When a part is cut free from the sheet material sometimes it's position shifts a little bit.  If holes are cut second then they can be slightly misaligned from their correct location.

<a href="https://imgur.com/S5U0eLm"><img src="https://i.imgur.com/S5U0eLm.jpg" title="source: imgur.com" /></a>

SVG file of above image:   https://pastebin.com/4dfXWb0c

If Deepnest could recognize interior lines (holes) and color these with a different stroke color when exporting SVG files then it would make if very easy to select these (select all lines with same stroke color), and be able to tell the laser cutter to cut those first.  Ideally there would be an option to set the stroke color for interior and exterior lines.

**Comment by deece** [comment]
What laser cutting software are you using? Both RDWorks & Lightburn automatically recognise internal cuts & do them first.

**Comment by napter** [comment]
This feature would be helpful to me too. In order to cut holes first on a glowforge laser you have to import an SVG with the interior cuts (holes) a different color than the exterior cuts.

**Comment by marmilicious** [comment]
@deece  We are using a Glowforge laser.   As @napter noted, the Glowforge needs to have interior lines be a different color in order to recognize and be able to cut those first.  But I could see this option being useful to others not using software as awesome as Lightburn.
(Lightburn really is great! Unfortunately we can't use it with the Glowforge.)

**Comment by deece** [comment]
It's probably worth pinging Glowforge to add this feature then, it's obvious and essential, and you shouldn't have to rely on third party applications to pre-process the data.

**Comment by napter** [comment]
At the very least I think it would make sense for Deepnest to leave the original color of any path, currently it removes the colors and sets them all to black.

**Comment by deece** [comment]
This is true. I do find that I get colours out when exporting as SVG (rather than DXF), although sometimes it is lost.

**Comment by marmilicious** [comment]
@napter  I agree, it would be good if input has color that it be preserved.  In our case we are using Onshape to design parts.  Onshape exports parts as a single color though, thus the wish that Deepnest might have an option to color holes differently on export and streamline the process a bit.  Anything to help simplify the work flow is appreciated.

**Comment by deece** [comment]
@marmilicious @napter Have you raised this with Glowforge? It's a feature common in many other laser cutting packages (even Chinese ones), and with a few people asking, they might make it a priority.

**Comment by napter** [comment]
@deece @marmilicious Is glowforge the only laser that allows you to change job settings based on color? A quick google shows the option to do color mapping with [Epilogs](https://www.epiloglaser.com/laser-machines/legend-laser/legend-color-mapping.htm). 

Regardless, if mapping cuts to colors isn't something to consider - keeping the original colors rather than losing them would be helpful. Thanks!

**Comment by marmilicious** [comment]
@napter  Yes, Epilog has the ability to choose presets based on color.  Keeping original colors (as set in design software) could certainly be useful to those users.

**Comment by deece** [comment]
@napter Most do, which is why the software should preserve colours. Any feature which deliberately alters the colours should be off by default.

Identification & sequencing of inner cuts is something that your CAM software is responsible for, so if it's missing, you should raise it there.

At the moment, it seems like @Jack000 has gone walkabout, so not only are there no patches to Deepnest coming in from him, but PRs are not getting merged either. If you want action, you'll need to look elsewhere.

**Comment by zaped212** [comment]
I would love to have this feature as well. I have a full spectrum laser and it uses colors to set speed, power, cut order, and what should be raster engraved.

It would be great if each level of nesting could be given its own unique color so parts ( with holes ) nested inside other parts would have 4 different colors one for each interior line.

**Comment by PowPodg** [comment]
One of the reasons that the holes are not recognized  is shown in #59

**Comment by rschenk** [comment]
I would love this feature as well—Deepnest preserves the original color of shapes when exporting as SVG. 

@deece you [mentioned](https://github.com/Jack000/Deepnest/issues/5#issuecomment-443371097) that some laser cam packages can automagically detect internal holes/features and sequence them first. That's fine if you only have internal holes that should be cut with the same settings as the outer perimeter, but supposing you wanted to also etch labels for those internal holes. Now you need a different power setting for the etch.  I set all this up with with different stroke colors in my design, but unfortunately Deepnest removes the colors and exports everything as black. It's a big pain in the neck to go back through and re-color the etches.

Edit: I think I misunderstood the original intent of this feature request, and @deece's reply that I linked above. It seems the original request is to have Deepnest perform the coloring by itself. For my use-case, it would suffice to simply preserve the original color that I've already specified in my input files, which is actually #99. My apologies for the confusion.

**Comment by jestelle** [comment]
Me too! Unfortunately Jack isn't working on this project right now. But I'm interested in reviving it. Any developers out there want to help? See #123
